src/Templates/AbstractTemplateCompatibility.php

```diff
@@ -0,0 +1,157 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Templates;
+
+/**
+ * BlockTemplatesCompatibility class.
+ *
+ * To bridge the gap on compatibility with PHP hooks and blockified templates.
+ *
+ * @internal
+ */
+abstract class AbstractTemplateCompatibility {
+	/**
+	 * The data of supported hooks, containing the hook name, the block name,
+	 * position, and the callbacks.
+	 *
+	 * @var array $hook_data The hook data.
+	 */
+	protected $hook_data;
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->set_hook_data();
+		$this->init();
+	}
+
+	/**
+	 * Initialization method.
+	 */
+	protected function init() {
+		if ( ! wc_current_theme_is_fse_theme() ) {
+			return;
+		}
+
+		add_filter( 'render_block_data', array( $this, 'update_render_block_data' ), 10, 3 );
+		add_filter( 'render_block', array( $this, 'inject_hooks' ), 10, 2 );
+	}
+
+	/**
+	 * Update the render block data to inject our custom attribute needed to
+	 * determine which blocks belong to an inherited Products block.
+	 *
+	 * @param array         $parsed_block The block being rendered.
+	 * @param array         $source_block An un-modified copy of $parsed_block, as it appeared in the source content.
+	 * @param WP_Block|null $parent_block If this is a nested block, a reference to the parent block.
+	 *
+	 * @return array
+	 */
+	abstract public function update_render_block_data( $parsed_block, $source_block, $parent_block );
+
+	/**
+	 * Inject hooks to rendered content of corresponding blocks.
+	 *
+	 * @param mixed $block_content The rendered block content.
+	 * @param mixed $block         The parsed block data.
+	 * @return string
+	 */
+	abstract public function inject_hooks( $block_content, $block );
+
+	/**
+	 * The hook data to inject to the rendered content of blocks. This also
+	 * contains hooked functions that will be removed by remove_default_hooks.
+	 *
+	 * The array format:
+	 * [
+	 *   <hook-name> => [
+	 *     block_name => <block-name>,
+	 *     position => before|after,
+	 *     hooked => [
+	 *       <function-name> => <priority>,
+	 *        ...
+	 *     ],
+	 *  ],
+	 * ]
+	 * Where:
+	 * - hook-name is the name of the hook that will be replaced.
+	 * - block-name is the name of the block that will replace the hook.
+	 * - position is the position of the block relative to the hook.
+	 * - hooked is an array of functions hooked to the hook that will be
+	 *   replaced. The key is the function name and the value is the
+	 *   priority.
+	 */
+	abstract protected function set_hook_data();
+
+
+	/**
+	 * Remove the default callback added by WooCommerce. We replaced these
+	 * callbacks by blocks so we have to remove them to prevent duplicated
+	 * content.
+	 */
+	protected function remove_default_hooks() {
+		foreach ( $this->hook_data as $hook => $data ) {
+			if ( ! isset( $data['hooked'] ) ) {
+				continue;
+			}
+			foreach ( $data['hooked'] as $callback => $priority ) {
+				remove_action( $hook, $callback, $priority );
```

Comment:
Is it correct to assume this is a temporary measure until we stop relying on the backend for rendering these blocks? I can imagine performance won't be optimal until we reach this point, considering we are removing the functions attached to these hooks to re-add them next.

Reply:
I'm tagging @dinhtungdu that he writes this part of the code.

> Is it correct to assume this is a temporary measure until we stop relying on the backend for rendering these blocks?

We will keep relying on the backend to render these blocks. It would be interesting to see how much overhead we are adding with this compatibility layer, but the real long-term is to remove this latter 🚀

Reply:
> We will keep relying on the backend to render these blocks. It would be interesting to see how much overhead we are adding with this compatibility layer, but the real long-term is to remove this latter 🚀

Cool! I asked because if we are planning to rely on this in the long term, we could consider making some changes to the core of Woo, so these functions are not hooked in the first place if the template in use is one of the new options we are introducing here.

Reply:
> Is it correct to assume this is a temporary measure until we stop relying on the backend for rendering these blocks?

Not really. It will be there even after we stop relying on the backend rendering for those blocks. We will remove the whole layer when the extensions ecosystem is ready for block themes.

> I can imagine performance won't be optimal until we reach this point, considering we are removing the functions attached to these hooks to re-add them next

@nefeline we won't re-add them later. We only remove default hooks added by WC Core because we have equivalent blocks in the templates, so there is no need to add them back. Btw, in terms of performance, I think removing hooks is pretty cheap.

> I asked because if we are planning to rely on this in the long term, we could consider making some changes to the core of Woo, so these functions are not hooked in the first place if the template in use is one of the new options we are introducing here.

We can always update WC Blocks when WC Core changes its default hooks handling.

Reply:
> Not really. It will be there even after we stop relying on the backend rendering for those blocks. We will remove the whole layer when the extensions ecosystem is ready for block themes.

Gotcha! Thanks for clarifying.

> we won't re-add them later. We only remove default hooks added by WC Core because we have equivalent blocks in the templates, so there is no need to add them back. Btw, in terms of performance, I think removing hooks is pretty cheap.

Yep, that's what I meant with re-adding them later: because in most cases (if not all of them), we are relying on the same template functions as the core templates are on Woo, so when a block is inserted, we are hooking the same methods/functions back.

Reply:
> so when a block is inserted, we are hooking the same methods/functions back.

Can you clarify your point more? I'm not sure I get this right.

We only prepend/append the hook (`do_action` calls) to the output of some blocks, we don't 'hook' any default methods added by WC Core (that are removed) back. Most of our blocks for archive/single product templates are now powered by WC template functions, but we use those functions directly in the render method of blocks, not hooking them back.

Reply:
> Can you clarify your point more? I'm not sure I get this right.

Sure! Taking the breadcrumbs block as an example. If we follow the order of execution, we have:

1. First, the core of Woo hooks `woocommerce_breadcrumb` to `'woocommerce_before_main_content'`:

[`add_action( 'woocommerce_before_main_content', 'woocommerce_breadcrumb', 20, 0 );`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-hooks.php#L47)

2. Next, if the breadcrumb block is used, we unhook this function via [`remove_default_hooks`](https://github.com/woocommerce/woocommerce-blocks/blob/ce9bd9cc9c676b8e5d191e7975133b530f322e4f/src/Templates/AbstractTemplateCompatibility.php#L92).

3. Last but not least, the [`woocommerce_breadcrumb`](https://github.com/woocommerce/woocommerce-blocks/blob/49ce2401047d87efdc7b6a3cbc516e3de70386f8/src/BlockTypes/Breadcrumbs.php#L30) function is then invoked by the block.

What I was wondering is if we can add a conditional on step 1 (something as `is_block_in_use( 'breadcrumb' )`, in the core of Woo) and remove the need for step 2 altogether.

Reply:
Actually, as a second thought: I believe this approach makes more sense only when Woo blocks start being the norm, not the exception (in other words, when the majority of the online stores are using blocks instead of classic templates): adding an `if` condition on core at this point might be a bit of an overhead. Marking as resolved: thanks for the discussion!